File: stripe/src/main/java/com/stripe/android/view/CountryAutoCompleteTextView.java
package com.stripe.android.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.annotation.VisibleForTesting;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.FrameLayout;

import com.stripe.android.R;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class CountryAutoCompleteTextView extends FrameLayout {
    @NonNull private final AutoCompleteTextView mCountryAutocomplete;
    @NonNull private final Map<String, String> mCountryNameToCode;

    @VisibleForTesting String mCountrySelected;
    @Nullable private CountryChangeListener mCountryChangeListener;

    public CountryAutoCompleteTextView(@NonNull Context context) {
        this(context, null);
    }

    public CountryAutoCompleteTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CountryAutoCompleteTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        inflate(getContext(), R.layout.country_autocomplete_textview, this);
        mCountryAutocomplete = findViewById(R.id.autocomplete_country_cat);
        mCountryNameToCode = CountryUtils.getCountryNameToCodeMap();
        final ArrayAdapter countryAdapter = new CountryAdapter(getContext(),
                new ArrayList<>(mCountryNameToCode.keySet()));
        mCountryAutocomplete.setThreshold(0);
        mCountryAutocomplete.setAdapter(countryAdapter);
        mCountryAutocomplete.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                final String countryEntered = mCountryAutocomplete.getText().toString();
                updateUiForCountryEntered(countryEntered);
            }
        });
        final String defaultCountryEntered =
                (String) Objects.requireNonNull(countryAdapter.getItem(0));
        updateUiForCountryEntered(defaultCountryEntered);
        mCountryAutocomplete.setText(defaultCountryEntered);
        mCountryAutocomplete.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean focused) {
                final String countryEntered = mCountryAutocomplete.getText().toString();
                if (focused) {
                    mCountryAutocomplete.showDropDown();
                } else {
                    updateUiForCountryEntered(countryEntered);
                }
            }
        });
    }

    /**
     * @return 2 digit country code of the country selected by this input.
     */
    @Nullable
    String getSelectedCountryCode() {
        return mCountrySelected;
    }

    /**
     * @param countryCode specify a country code to display in the input. The input will display
     *                    the full country display name.
     */
    void setCountrySelected(@Nullable String countryCode) {
        if (countryCode == null) {
            return;
        }
        updateUiForCountryEntered(getDisplayCountry(countryCode));
    }

    void setCountryChangeListener(@Nullable CountryChangeListener countryChangeListener) {
        mCountryChangeListener = countryChangeListener;
    }

    @VisibleForTesting
    void updateUiForCountryEntered(@NonNull String displayCountryEntered) {
        final String countryCodeEntered = mCountryNameToCode.get(displayCountryEntered);
        if (countryCodeEntered != null) {
            if (mCountrySelected == null || !mCountrySelected.equals(countryCodeEntered)) {
                mCountrySelected = countryCodeEntered;
                if (mCountryChangeListener != null) {
                    mCountryChangeListener.onCountryChanged(mCountrySelected);
                }
            }
            mCountryAutocomplete.setText(displayCountryEntered);
        } else if (mCountrySelected != null) {
            // Revert back to last valid country if country is not recognized.
            mCountryAutocomplete.setText(getDisplayCountry(mCountrySelected));
        }
    }

    @NonNull
    private static String getDisplayCountry(@NonNull String countryCode) {
        return new Locale("", countryCode).getDisplayCountry();
    }

    interface CountryChangeListener {
        void onCountryChanged(@NonNull String countryCode);
    }
}

File: stripe/src/main/java/com/stripe/android/view/CountryAutoCompleteTextView.kt
package com.stripe.android.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View.OnFocusChangeListener
import android.widget.AdapterView
import android.widget.AutoCompleteTextView
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.annotation.StyleRes
import androidx.annotation.VisibleForTesting
import androidx.core.content.withStyledAttributes
import androidx.core.os.ConfigurationCompat
import com.google.android.material.textfield.TextInputLayout
import com.stripe.android.R

/**
 * A subclass of [TextInputLayout] that programmatically wraps a styleable [AutoCompleteTextView],
 * which configures a [CountryAdapter] to display list of countries in its popup.
 *
 * The style of [AutoCompleteTextView] can be changed via [R.styleable.CountryAutoCompleteTextView_countryAutoCompleteStyle],
 * the style of popup items can be changed via [R.styleable.CountryAutoCompleteTextView_countryItemLayout].
 */
internal class CountryAutoCompleteTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = com.google.android.material.R.attr.textInputStyle
) : TextInputLayout(context, attrs, defStyleAttr) {

    @StyleRes
    private var countryAutoCompleteStyleRes: Int = INVALID_COUNTRY_AUTO_COMPLETE_STYLE

    @LayoutRes
    private var countryLineLayoutRes: Int = DEFAULT_COUNTRY_LINE_LAYOUT

    @VisibleForTesting
    internal val countryAutocomplete: AutoCompleteTextView

    /**
     * The 2 digit country code of the country selected by this input.
     */
    @VisibleForTesting
    var selectedCountry: Country? = null
        set(value) {
            field = value
            value?.let {
                countryChangeCallback(it)
            }
        }

    @JvmSynthetic
    internal var countryChangeCallback: (Country) -> Unit = {}

    private var countryAdapter: CountryAdapter

    init {
        context.withStyledAttributes(
            attrs,
            R.styleable.CountryAutoCompleteTextView
        ) {
            countryAutoCompleteStyleRes = getResourceId(
                R.styleable.CountryAutoCompleteTextView_countryAutoCompleteStyle,
                INVALID_COUNTRY_AUTO_COMPLETE_STYLE
            )

            countryLineLayoutRes = getResourceId(
                R.styleable.CountryAutoCompleteTextView_countryItemLayout,
                DEFAULT_COUNTRY_LINE_LAYOUT
            )
        }

        countryAutocomplete = AutoCompleteTextView(
            context,
            null,
            R.attr.autoCompleteTextViewStyle
        ).takeIf { countryAutoCompleteStyleRes == INVALID_COUNTRY_AUTO_COMPLETE_STYLE }
            ?: AutoCompleteTextView(context, null, 0, countryAutoCompleteStyleRes)

        addView(
            countryAutocomplete,
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        )

        LayoutInflater.from(context).let { layoutInflater ->
            countryAdapter = CountryAdapter(
                context,
                CountryUtils.getOrderedCountries(
                    ConfigurationCompat.getLocales(context.resources.configuration)[0]
                ),
                countryLineLayoutRes
            ) {
                // document must be a text view
                layoutInflater.inflate(countryLineLayoutRes, it, false) as TextView
            }
        }

        countryAutocomplete.threshold = 0
        countryAutocomplete.setAdapter(countryAdapter)
        countryAutocomplete.onItemClickListener =
            AdapterView.OnItemClickListener { _, _, position, _ ->
                updatedSelectedCountryCode(countryAdapter.getItem(position))
            }
        countryAutocomplete.onFocusChangeListener = OnFocusChangeListener { _, focused ->
            if (focused) {
                countryAutocomplete.showDropDown()
            } else {
                val countryEntered = countryAutocomplete.text.toString()
                updateUiForCountryEntered(countryEntered)
            }
        }

        selectedCountry = countryAdapter.firstItem
        updateInitialCountry()

        val errorMessage = resources.getString(R.string.address_country_invalid)

        countryAutocomplete.validator = CountryAutoCompleteTextViewValidator(
            countryAdapter
        ) { country ->
            selectedCountry = country

            if (country != null) {
                clearError()
            } else {
                error = errorMessage
                isErrorEnabled = true
            }
        }
    }

    private fun updateInitialCountry() {
        val initialCountry = countryAdapter.firstItem
        countryAutocomplete.setText(initialCountry.name)
        selectedCountry = initialCountry
    }

    /**
     * @param allowedCountryCodes A set of allowed country codes. Will be ignored if empty.
     */
    internal fun setAllowedCountryCodes(allowedCountryCodes: Set<String>) {
        val isUpdated = countryAdapter.updateUnfilteredCountries(allowedCountryCodes)
        if (isUpdated) {
            updateInitialCountry()
        }
    }

    /**
     * @param countryCode specify a country code to display in the input. The input will display
     * the full country display name.
     */
    internal fun setCountrySelected(countryCode: String) {
        updateUiForCountryEntered(CountryUtils.getDisplayCountry(countryCode))
    }

    @VisibleForTesting
    internal fun updateUiForCountryEntered(displayCountryEntered: String) {
        val country = CountryUtils.getCountryByName(displayCountryEntered)

        // If the user-typed country matches a valid country, update the selected country
        // Otherwise, revert back to last valid country if country is not recognized.
        val displayCountry = country?.let {
            updatedSelectedCountryCode(it)
            displayCountryEntered
        } ?: selectedCountry?.name

        countryAutocomplete.setText(displayCountry)
    }

    private fun updatedSelectedCountryCode(country: Country) {
        clearError()
        if (selectedCountry != country) {
            selectedCountry = country
        }
    }

    internal fun validateCountry() {
        countryAutocomplete.performValidation()
    }

    private fun clearError() {
        error = null
        isErrorEnabled = false
    }

    private companion object {
        const val INVALID_COUNTRY_AUTO_COMPLETE_STYLE = 0
        val DEFAULT_COUNTRY_LINE_LAYOUT = R.layout.country_text_view
    }
}
